code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm

```diff
@@ -1,261 +1,357 @@
 
-	if(X.agility)
-		to_chat(X, span_xenowarning("We lower ourselves to all fours and loosen our armored scales to ease our movement."))
-		X.add_movespeed_modifier(MOVESPEED_ID_WARRIOR_AGILITY , TRUE, 0, NONE, TRUE, X.xeno_caste.agility_speed_increase)
-		X.soft_armor = X.soft_armor.modifyAllRatings(last_agility_bonus)
-		owner.toggle_move_intent(MOVE_INTENT_RUN) //By default we swap to running when activating agility
-	else
-		to_chat(X, span_xenowarning("We raise ourselves to stand on two feet, hard scales setting back into place."))
-		X.remove_movespeed_modifier(MOVESPEED_ID_WARRIOR_AGILITY)
-		X.soft_armor = X.soft_armor.modifyAllRatings(-last_agility_bonus)
+/// Adds an outline around the ability button to represent Empower.
+/datum/action/ability/activable/xeno/warrior/proc/add_empowered_frame()
+	button.add_overlay(visual_references[VREF_MUTABLE_EMPOWERED_FRAME])
+
+/// Removes the Empower outline.
+/datum/action/ability/activable/xeno/warrior/proc/remove_empowered_frame()
+	button.cut_overlay(visual_references[VREF_MUTABLE_EMPOWERED_FRAME])
+
+/// Handles anything that would happen when a target is thrown into an atom using an ability.
+/datum/action/ability/activable/xeno/warrior/proc/thrown_into(datum/source, atom/hit_atom, impact_speed)
+	SIGNAL_HANDLER
+	UnregisterSignal(source, COMSIG_MOVABLE_IMPACT)
+	var/mob/living/living_target = source
+	INVOKE_ASYNC(living_target, TYPE_PROC_REF(/mob, emote), "pain")
+	living_target.Knockdown(WARRIOR_DISPLACE_KNOCKDOWN)
+	var/mob/living/carbon/xenomorph/xeno_owner = owner
+	new /obj/effect/temp_visual/warrior/impact(get_turf(living_target), get_dir(living_target, xeno_owner))
+	// mob/living/turf_collision() does speed * 5 damage on impact with a turf, and we don't want to go overboard, so we deduce that here.
+	var/thrown_damage = (xeno_owner.xeno_caste.melee_damage * xeno_owner.xeno_melee_damage_modifier) - (impact_speed * 5)
+	living_target.apply_damage(thrown_damage, BRUTE, blocked = MELEE)
+	if(isliving(hit_atom))
+		var/mob/living/hit_living = hit_atom
+		if(hit_living.issamexenohive(xeno_owner))
+			return
+		INVOKE_ASYNC(hit_living, TYPE_PROC_REF(/mob, emote), "pain")
+		hit_living.apply_damage(thrown_damage, BRUTE, blocked = MELEE)
+		hit_living.Knockdown(WARRIOR_DISPLACE_KNOCKDOWN)
+		step_away(hit_living, living_target, 1, 1)
+	if(isobj(hit_atom))
+		var/obj/hit_object = hit_atom
+		if(istype(hit_object, /obj/structure/xeno))
+			return
+		hit_object.take_damage(thrown_damage, BRUTE)
+	if(iswallturf(hit_atom))
+		var/turf/closed/wall/hit_wall = hit_atom
+		if(!hit_wall.resistance_flags != INDESTRUCTIBLE)
+			hit_wall.take_damage(thrown_damage, BRUTE)
+
+/// Ends the target's throw.
+/datum/action/ability/activable/xeno/warrior/proc/throw_ended(datum/source)
+	SIGNAL_HANDLER
+	UnregisterSignal(source, COMSIG_MOVABLE_POST_THROW)
+	addtimer(CALLBACK(src, TYPE_PROC_REF(/datum, UnregisterSignal), source, COMSIG_MOVABLE_IMPACT, COMSIG_MOVABLE_POST_THROW), 1) // having a giggle
```

Comment:
I put a comment above this line of code explaining it, but I'm gonna put it here as well:
- So the reason why we do not flat out unregister this is because, when an atom makes impact with something, it calls throw_impact(). Calling it this way causes stop_throw() to be called in most cases, because impacts can cause a bounce effect and ending the throw makes it happen. Given the way we have signals setup, unregistering it at that point would cause thrown_into() to never get called, and that is exactly the reason why the line of code below exists.